### PR TITLE
DRV-373: migrate HttpClient to HTTP2 protocol for NodeJS environment

### PR DESCRIPTION
**Pull request by tigger9flow**
### Notes
[Jira Ticket](https://faunadb.atlassian.net/browse/DRV-373)
The following changes introduce migration to HTTP2 protocol for NodeJS environment. All HTTP-related logic has been extracted and encapsulated in the `HttpClient` class. Environment-specific protocol mechanics are handled in adapters that expose the same interface with the `.execute` method. Thus there are `FetchAdapter` based on Fetch API and `Http2Adapter` backed by native `http2` NodeJS module. `Http2Adapter` is applicable only when it's NodeJS environment and no Fetch API override has been provided to preserve backward-compatibility.
`TimeoutError` and `AbortError` error classes have been added to be thrown by both adapters to expose the same API for handling aborts/timeouts.
Currently, there's no `onEnd` event exposed from the `HttpClient`, if a stream is ended for some reason, the `onError` hook is triggered with `TypeError('network error')` error instead to align on how it works in a browser environment.

### How to test
No new features have been introduced. Can be tested regression-wise.

### CODE REVIEW

File: src/_http/index.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    method: options.method || 'GET',
+    headers: util.removeNullAndUndefinedValues(headers),
+    body: options.body,
+    signal: options.signal,
+    timeout: this._timeout,
+    streamConsumer: options.streamConsumer,
+  })
+}
+
+/** @ignore */
```
Comment (parkhomenko) [inline src/_http/index.js:118]: Is this annotation used to ignore test coverage for this function? We could just have a test for it.

File: src/_http/index.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  var invalidStreamConsumer =
+    options.streamConsumer &&
+    (typeof options.streamConsumer.onData !== 'function' ||
+      typeof options.streamConsumer.onError !== 'function')
+
+  if (invalidStreamConsumer) {
+    return Promise.reject(new TypeError('Invalid "streamConsumer" provided'))
+  }
+
+  var url = parse(this._baseUrl)
```
Comment (parkhomenko) [inline src/_http/index.js:94]: There were some changes to decrease bundle size. One of them was a removal of the `parse` function. You might want to update this:
https://github.com/fauna/faunadb-js/commit/835feceb6d5647a29b79359b369a752665598fb9

**Review by parkhomenko — CHANGES_REQUESTED** [review]
Great work!
There is only a comment about including the changes of a recent PR about bundle size decrease.

**Review by parkhomenko — DISMISSED** [review]

**Review by fireridlle — APPROVED** [review]

**Review by parkhomenko — APPROVED** [review]

**Review by fireridlle — DISMISSED** [review]

File: src/_http/errors.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  this.message = message || 'Request aborted'
+  this.isAbortError = true
+}
+
+util.inherits(AbortError, Error)
+
+module.exports = {
+  TimeoutError: TimeoutError,
+  AbortError: AbortError,
+}
```
Comment (erickpintor) [inline src/_http/errors.js:39]: Q: Are these new or were they existing errors?

File: src/_http/fetchAdapter.js [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+ * Http client adapter built around fetch API.
+ *
+ * @constructor
+ * @param {?object} options FetchAdapter options.
+ * @param {?boolean} options.keepAlive Whether use keep-alive connection.
+ * @param {?boolean} options.isHttps Whether use https connection.
+ * @param {?function} options.fetch Fetch compatible API.
+ * @private
+ */
+function FetchAdapter(options) {
```
Comment (erickpintor) [inline src/_http/fetchAdapter.js:17]: A much needed refactor! I appreciate that.

File: src/_http/http2Adapter.js [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    // Initializing http2 session.
+    this._sessionMap[sessionKey] = http2
+      .connect(origin)
+      .once('error', cleanup)
+      .once('goaway', cleanup)
+      // Destroys http2 session after specified time of inactivity
+      // and releases event loop.
+      .setTimeout(DESTROY_HTTP2_SESSION_TIME, cleanup)
```
Comment (erickpintor) [inline src/_http/http2Adapter.js:58]: Q: what does "inactivity" mean? Is it lack of network traffic or lack of open open streams?

File: src/_http/errors.js (hunk 3 above)
Comment (tigger9flow) [reply src/_http/errors.js:39]: These are new errors shared across both adapters, to expose the same error API from `HttpClient` outside. E.g. [here](https://github.com/fauna/faunadb-js/pull/396/files#diff-a36fdfce533ef260dcd9afc30b77d10eecb0c6e2711fbcbef70e66a1ca5ab7d5R133) we assert `AbortError` no matter what underlying adapter is used as they share the same errors for abort / timeout

Comment (tigger9flow) [reply src/_http/errors.js:39]: These are new errors shared across both adapters, to expose the same error API from `HttpClient` outside. E.g. [here](https://github.com/fauna/faunadb-js/pull/396/files#diff-a36fdfce533ef260dcd9afc30b77d10eecb0c6e2711fbcbef70e66a1ca5ab7d5R133) we assert `AbortError` no matter what underlying adapter is used as they share the same errors for abort / timeout

File: src/_http/http2Adapter.js (hunk 5 above)
Comment (tigger9flow) [reply src/_http/http2Adapter.js:58]: Yeah, it's a lack of any network traffic (including non-streaming requests)

Comment (erickpintor) [reply src/_http/http2Adapter.js:58]: I mean streams in the HTTP2 spec sense. Would this cause idle **fauna streams** to close after the section timeout?

Comment (tigger9flow) [reply src/_http/http2Adapter.js:58]: No, it wouldn't. When there's an ongoing http2 stream in the session, it won't be closed

**Review by erickpintor — APPROVED** [review]
LGTM

**Review by parkhomenko — APPROVED** [review]

**Review by fireridlle — APPROVED** [review]

**Pull request merged**